File: DI-Template-Training/src/blob_storage.py
"""Azure Blob Storage utilities for training data management."""

import os
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Optional

from azure.storage.blob import (
    BlobServiceClient,
    ContainerSasPermissions,
    generate_container_sas,
)

from .utils import print_progress


def ensure_container_exists(
    client: BlobServiceClient,
    container_name: str,
) -> bool:
    """
    Ensure a blob container exists, creating it if necessary.

    Args:
        client: BlobServiceClient instance
        container_name: Name of the container

    Returns:
        True if container was created, False if it already existed
    """
    container_client = client.get_container_client(container_name)
    try:
        container_client.get_container_properties()
        return False  # Already exists
    except Exception:
        container_client.create_container()
        return True  # Created


def upload_file_to_blob(
    client: BlobServiceClient,
    container_name: str,
    local_path: Path | str,
    blob_name: Optional[str] = None,
    overwrite: bool = True,
) -> str:
    """
    Upload a single file to blob storage.

    Args:
        client: BlobServiceClient instance
        container_name: Target container name
        local_path: Local file path
        blob_name: Name in blob storage (defaults to filename)
        overwrite: Whether to overwrite existing blobs

    Returns:
        Blob URL
    """
    local_path = Path(local_path)
    blob_name = blob_name or local_path.name

    container_client = client.get_container_client(container_name)
    blob_client = container_client.get_blob_client(blob_name)

    with open(local_path, "rb") as f:
        blob_client.upload_blob(f, overwrite=overwrite)

    return blob_client.url


def upload_training_data(
    client: BlobServiceClient,
    container_name: str,
    training_dir: Path | str,
    verbose: bool = True,
) -> dict:
    """
    Upload all training data files to blob storage.

    Expected files in training_dir:
    - fields.json
    - *.jpg (training images)
    - *.jpg.ocr.json (OCR output for each image)
    - *.jpg.labels.json (labels for each image)

    Args:
        client: BlobServiceClient instance
        container_name: Target container name
        training_dir: Directory containing training data
        verbose: Print progress

    Returns:
        Dict with upload statistics
    """
    training_dir = Path(training_dir)

    # Ensure container exists
    created = ensure_container_exists(client, container_name)
    if verbose and created:
        print(f"Created container: {container_name}")

    # Find all files to upload
    files_to_upload = []

    # fields.json (required)
    fields_json = training_dir / "fields.json"
    if fields_json.exists():
        files_to_upload.append(fields_json)
    else:
        raise FileNotFoundError(f"fields.json not found in {training_dir}")

    # Training images and their associated files
    for img_path in training_dir.glob("*.jpg"):
        files_to_upload.append(img_path)

        # OCR JSON
        ocr_path = img_path.parent / f"{img_path.name}.ocr.json"
        if ocr_path.exists():
            files_to_upload.append(ocr_path)

        # Labels JSON
        labels_path = img_path.parent / f"{img_path.name}.labels.json"
        if labels_path.exists():
            files_to_upload.append(labels_path)

    # Upload all files
    uploaded = []
    failed = []

    for i, file_path in enumerate(files_to_upload):
        if verbose:
            print_progress(i + 1, len(files_to_upload), "Uploading")

        try:
            url = upload_file_to_blob(client, container_name, file_path)
            uploaded.append({"file": file_path.name, "url": url})
        except Exception as e:
            failed.append({"file": file_path.name, "error": str(e)})

    return {
        "container": container_name,
        "total_files": len(files_to_upload),
        "uploaded": len(uploaded),
        "failed": len(failed),
        "uploaded_files": uploaded,
        "failed_files": failed,
    }


def generate_sas_url(
    client: BlobServiceClient,
    container_name: str,
    expiry_days: int = 7,
) -> str:
    """
    Generate a SAS URL for a container.

    Args:
        client: BlobServiceClient instance
        container_name: Container name
        expiry_days: Number of days until SAS expires

    Returns:
        Full container URL with SAS token
    """
    # Get account details from client
    account_name = client.account_name
    account_key = client.credential.account_key

    # Generate SAS token
    sas_token = generate_container_sas(
        account_name=account_name,
        container_name=container_name,
        account_key=account_key,
        permission=ContainerSasPermissions(read=True, list=True),
        expiry=datetime.now(timezone.utc) + timedelta(days=expiry_days),
    )

    # Build full URL
    container_client = client.get_container_client(container_name)
    return f"{container_client.url}?{sas_token}"


def list_container_blobs(
    client: BlobServiceClient,
    container_name: str,
    prefix: Optional[str] = None,
) -> list[dict]:
    """
    List blobs in a container.

    Args:
        client: BlobServiceClient instance
        container_name: Container name
        prefix: Optional prefix filter

    Returns:
        List of blob info dicts
    """
    container_client = client.get_container_client(container_name)

    blobs = []
    for blob in container_client.list_blobs(name_starts_with=prefix):
        blobs.append({
            "name": blob.name,
            "size": blob.size,
            "last_modified": blob.last_modified,
            "content_type": blob.content_settings.content_type if blob.content_settings else None,
        })

    return blobs


def delete_container_contents(
    client: BlobServiceClient,
    container_name: str,
    prefix: Optional[str] = None,
    verbose: bool = True,
) -> int:
    """
    Delete all blobs in a container (or with a prefix).

    Args:
        client: BlobServiceClient instance
        container_name: Container name
        prefix: Optional prefix to filter blobs
        verbose: Print progress

    Returns:
        Number of blobs deleted
    """
    container_client = client.get_container_client(container_name)

    blobs = list(container_client.list_blobs(name_starts_with=prefix))

    for i, blob in enumerate(blobs):
        if verbose:
            print_progress(i + 1, len(blobs), "Deleting")
        container_client.delete_blob(blob.name)

    return len(blobs)

File: Document-Classification/Azure-Document-Intelligence/supporting-code/blobStorage.ts
import {
  BlobServiceClient,
  ContainerClient,
  ContainerSASPermissions,
  generateBlobSASQueryParameters,
  SASProtocol,
  StorageSharedKeyCredential,
} from '@azure/storage-blob';
import { Buffer } from "node:buffer";

export interface UploadFileResult {
  fileName: string;
  url: string;
  size: number;
}

export interface UploadResult {
  containerName: string;
  totalFiles: number;
  uploaded: number;
  failed: number;
  uploadedFiles: UploadFileResult[];
  failedFiles: { fileName: string; error: string }[];
}

export interface BlobInfo {
  name: string;
  size: number;
  lastModified: Date;
  contentType?: string;
}

export class BlobStorage {
  private blobServiceClient!: BlobServiceClient;
  private accountName: string;
  private accountKey: string;
  private readonly deleteRetryDelayMs = 5000;
  private readonly deleteRetryAttempts = 24;

  constructor() {
    const connectionString = Deno.env.get("AZURE_STORAGE_CONNECTION_STRING");
    this.accountName = Deno.env.get("AZURE_STORAGE_ACCOUNT_NAME") || '';
    this.accountKey = Deno.env.get("AZURE_STORAGE_ACCOUNT_KEY") || '';

    if (!connectionString) {
      console.warn(
        'AZURE_STORAGE_CONNECTION_STRING not configured. Blob storage features will not work.',
      );
      return;
    }

    this.blobServiceClient =
      BlobServiceClient.fromConnectionString(connectionString);
    console.log('Blob Storage client initialized');
  }

  /**
   * Ensure a container exists, creating it if necessary
   */
  async ensureContainerExists(containerName: string): Promise<boolean> {
    const containerClient =
      this.blobServiceClient.getContainerClient(containerName);

    for (let attempt = 1; attempt <= this.deleteRetryAttempts; attempt += 1) {
      try {
        // Try to create directly; this is the reliable signal for reuse.
        await containerClient.create();
        console.log(`Created container: ${containerName}`);
        return true;
      } catch (error: any) {
        const message = error.message || '';
        const statusCode = error.statusCode || error.status;
        const isAlreadyExists =
          statusCode === 409 &&
          (error.code === 'ContainerAlreadyExists' ||
            message.includes('ContainerAlreadyExists'));
        const isBeingDeleted =
          statusCode === 409 &&
          (error.code === 'ContainerBeingDeleted' ||
            message.includes('being deleted'));

        if (isAlreadyExists) {
          console.debug(`Container ${containerName} already exists`);
          return false;
        }

        if (isBeingDeleted && attempt < this.deleteRetryAttempts) {
          console.warn(
            `Container ${containerName} is being deleted. Retry ${attempt}/${this.deleteRetryAttempts}`,
          );
          await this.delay(this.deleteRetryDelayMs);
          continue;
        }

        console.error(
          `Failed to ensure container exists: ${containerName}`,
          error.stack,
        );
        throw error;
      }
    }

    throw new Error(
      `Failed to ensure container exists after ${this.deleteRetryAttempts} attempts: ${containerName}`,
    );
  }

  /**
   * Upload a single file to blob storage
   */
  async uploadFile(
    containerName: string,
    blobName: string,
    content: Buffer | string,
  ): Promise<string> {
    try {
      const containerClient =
        this.blobServiceClient.getContainerClient(containerName);
      const blockBlobClient = containerClient.getBlockBlobClient(blobName);

      const buffer = typeof content === 'string' ? Buffer.from(content) : content;

      await blockBlobClient.uploadData(buffer);

      console.debug(`Uploaded blob: ${blobName} to ${containerName}`);
      return blockBlobClient.url;
    } catch (error: any) {
      console.error(`Failed to upload blob: ${blobName}`, error.stack);
      throw error;
    }
  }

  /**
   * Upload multiple files to blob storage
   */
  async uploadFiles(
    containerName: string,
    files: { name: string; content: Buffer | string }[],
  ): Promise<UploadResult> {
    // Ensure container exists
    await this.ensureContainerExists(containerName);

    const uploadedFiles: UploadFileResult[] = [];
    const failedFiles: { fileName: string; error: string }[] = [];

    for (const file of files) {
      try {
        const url = await this.uploadFile(
          containerName,
          file.name,
          file.content,
        );
        const buffer =
          typeof file.content === 'string'
            ? Buffer.from(file.content)
            : file.content;

        uploadedFiles.push({
          fileName: file.name,
          url,
          size: buffer.length,
        });
      } catch (error: any) {
        failedFiles.push({
          fileName: file.name,
          error: error.message,
        });
      }
    }

    return {
      containerName,
      totalFiles: files.length,
      uploaded: uploadedFiles.length,
      failed: failedFiles.length,
      uploadedFiles,
      failedFiles,
    };
  }

  /**
   * Generate a SAS URL for a container with read and list permissions
   */
  async generateSasUrl(
    containerName: string,
    expiryDays = 7,
  ): Promise<string> {
    try {
      if (!this.accountName || !this.accountKey) {
        throw new Error('Azure Storage account credentials not configured');
      }

      const containerClient =
        this.blobServiceClient.getContainerClient(containerName);

      const sharedKeyCredential = new StorageSharedKeyCredential(
        this.accountName,
        this.accountKey,
      );

      const startsOn = new Date();
      startsOn.setMinutes(startsOn.getMinutes() - 5);
      const expiresOn = new Date();
      expiresOn.setDate(expiresOn.getDate() + expiryDays);

      const permissions = ContainerSASPermissions.parse('rl');

      const sasToken = generateBlobSASQueryParameters(
        {
          containerName,
          permissions,
          startsOn,
          expiresOn,
          protocol: SASProtocol.Https,
        },
        sharedKeyCredential,
      ).toString();

      const sasUrl = `${containerClient.url}?${sasToken}`;

      console.log(
        `Generated SAS URL for container ${containerName} (expires in ${expiryDays} days)`,
      );

      return sasUrl;
    } catch (error: any) {
      console.error(
        `Failed to generate SAS URL for container: ${containerName}`,
        error.stack,
      );
      throw error;
    }
  }



  /**
   * Validate a container SAS URL by attempting to list blobs.
   */
  async validateContainerSasUrl(sasUrl: string): Promise<{
    canList: boolean;
    error?: string;
    blobCount?: number;
    sampleBlobs?: string[];
  }> {
    try {
      const containerClient = new ContainerClient(sasUrl);
      const sampleBlobs: string[] = [];
      let count = 0;

      for await (const blob of containerClient.listBlobsFlat()) {
        count += 1;
        if (sampleBlobs.length < 5) {
          sampleBlobs.push(blob.name);
        }
      }

      return {
        canList: true,
        blobCount: count,
        sampleBlobs,
      };
    } catch (error: any) {
      return {
        canList: false,
        error: error.message,
      };
    }
  }

  /**
   * List all blobs in a container
   */
  async listBlobs(
    containerName: string,
    prefix?: string,
  ): Promise<BlobInfo[]> {
    try {
      const containerClient =
        this.blobServiceClient.getContainerClient(containerName);
      const blobs: BlobInfo[] = [];

      for await (const blob of containerClient.listBlobsFlat({
        prefix,
      })) {
        blobs.push({
          name: blob.name,
          size: blob.properties.contentLength || 0,
          lastModified: blob.properties.lastModified,
          contentType: blob.properties.contentType,
        });
      }

      console.debug(
        `Listed ${blobs.length} blobs in container ${containerName}`,
      );
      return blobs;
    } catch (error: any) {
      console.error(
        `Failed to list blobs in container: ${containerName}`,
        error.stack,
      );
      throw error;
    }
  }

  /**
   * Delete a container and all its contents
   */
  async deleteContainer(containerName: string): Promise<void> {
    try {
      const containerClient =
        this.blobServiceClient.getContainerClient(containerName);
      await containerClient.delete();
      console.log(`Deleted container: ${containerName}`);
    } catch (error: any) {
      console.error(
        `Failed to delete container: ${containerName}`,
        error.stack,
      );
      throw error;
    }
  }

  /**
   * Delete a container if it exists (ignore not found).
   */
  async deleteContainerIfExists(containerName: string): Promise<boolean> {
    try {
      const containerClient =
        this.blobServiceClient.getContainerClient(containerName);
      const result = await containerClient.deleteIfExists();
      if (result.succeeded) {
        console.log(`Deleted container: ${containerName}`);
      } else {
        console.debug(`Container not found: ${containerName}`);
        return false;
      }
      return true;
    } catch (error: any) {
      console.error(
        `Failed to delete container: ${containerName}`,
        error.stack,
      );
      throw error;
    }
  }

  /**
   * Delete all blobs in a container without removing the container itself.
   */
  async clearContainerContents(containerName: string): Promise<number> {
    try {
      await this.ensureContainerExists(containerName);
      const containerClient =
        this.blobServiceClient.getContainerClient(containerName);
      let deleted = 0;

      for await (const blob of containerClient.listBlobsFlat()) {
        await containerClient.deleteBlob(blob.name);
        deleted += 1;
      }

      console.log(
        `Cleared ${deleted} blob(s) from container: ${containerName}`,
      );
      return deleted;
    } catch (error: any) {
      console.error(
        `Failed to clear container contents: ${containerName}`,
        error.stack,
      );
      throw error;
    }
  }

  private async delay(ms: number): Promise<void> {
    await new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Get container client for advanced operations
   */
  getContainerClient(containerName: string): ContainerClient {
    return this.blobServiceClient.getContainerClient(containerName);
  }

  /**
 * Generate a SAS URL for a specific blob
 * @param containerName The name of the container
 * @param blobName The path/name of the blob
 * @param expiresInMinutes How long the SAS should be valid (default: 60)
 */
  getBlobSasUrl(containerName: string, blobName: string, expiresInMinutes = 60): string {
    if (!this.accountName || !this.accountKey) {
      throw new Error('Storage account name/key not configured.');
    }
    const sharedKeyCredential = new StorageSharedKeyCredential(this.accountName, this.accountKey);
    const containerClient = this.blobServiceClient.getContainerClient(containerName);
    const blobClient = containerClient.getBlobClient(blobName);
    const now = new Date();
    const expires = new Date(now.getTime() + expiresInMinutes * 60 * 1000);
    const sas = generateBlobSASQueryParameters({
      containerName,
      blobName,
      permissions: ContainerSASPermissions.parse('r'),
      startsOn: now,
      expiresOn: expires,
      protocol: SASProtocol.Https,
    }, sharedKeyCredential).toString();
    return `${blobClient.url}?${sas}`;
  }
}
